Add internal scripts to do Web QA specific stuff

bobsilverberg opened the pull request:
Fixes #2
Fixes #3
Fixes #9
Fixes #10

Adds the ability to trigger a meeting announcement via a HTTP GET, which could be used for scheduling

I'm going to create a Heroku review app for this so we can test it interactively. :)

bobsilverberg commented:
Nice, it seems to work! You can even test the HTTP GET thing by going to https://webqabot-pr-12.herokuapp.com/hubot/announce_meeting

@davehunt r?

BTW, the reason I am working on this is that I like the new GitHub activity notification stuff so much I really want to get this new version of the bot ready to replace our existing bot.

davehunt commented on hubot-scripts.json:
```diff
@@ -1 +1 @@
-[]
\ No newline at end of file
+["webqabot"]
```
I think I'd call this "webqa" to distinguish the bot from the script.

davehunt commented on scripts/webqabot.coffee:
```diff
@@ -0,0 +1,40 @@
+
+  robot.respond /lmgtfy (.*)/i, (res) ->
+    topic = encodeURIComponent(res.match[1]).replace(/[!'()*]/g, escape)
+    res.send "http://lmgtfy.com/?q=#{topic}"
```
Not creating an external script is fine, but I think this should be a separate internal script.

davehunt commented:
I'd prefer separate commits at least for lmgtfy and webqa scripts, but r+ wc

bobsilverberg commented:
@davehunt Thanks for the review. Tbh, I think lmgtfy is a bit gimmicky, and not really a particularly worthwhile script, so I'm not sure it should ever be converted to an external script. I don't really like the idea of creating a whole bunch of files with two-line scripts in them either, so I am actually against the idea of breaking this into separate scripts (I realize you are only suggesting breaking out lmgtfy right now). I am open to be convinced otherwise though.

bobsilverberg commented:
@davehunt I renamed the internal script and added some comments to it. I removed the part that responds to an HTTP GET because, while it was kind of cool to do, scheduling something to hit that url will probably be more complicated than just using hubot-cron to ask the bot to announce the meeting.

Still awaiting your response to my lmgtfy thing.

davehunt commented on scripts/webqa.coffee:
```diff
@@ -0,0 +1,41 @@
+    if announce?
+      meeting_time = "in 15 minutes for our team meeting"
+    else
+      meeting_time = "at 9AM Pacific every Thursday"
```
Could you provide some details on how the cron works? If possible, it might be nice to have an optional `when` argument. Then a cron could be set up to pass 'in 15 minutes' or other values. If it's not passed, we would default to 'at 9am Pacific every Thursday'. Also, I think we should include 'for our team meeting' in any case.

davehunt replied:
Ah, I see that hubot-cron basically just says something on a schedule. This means that anyone saying 'webqabot: !meeting' would get the same response, even if it's not true. We should be able to respond to something like 'webqabot meeting 15 minutes' by matching whatever is said after the meeting command. This might be overthinking it, as it does mean we'd be able to get the bot to repeat arbitrary things, but I guess hubot-cron essentially allows that too.

davehunt commented on scripts/webqa.coffee:
```diff
@@ -0,0 +1,41 @@
+  # respond to lmgtfy
+  robot.respond /lmgtfy (.*)/i, (res) ->
+    topic = encodeURIComponent(res.match[1]).replace(/[!'()*]/g, escape)
+    res.send "http://lmgtfy.com/?q=#{topic}"
```
I agree that this is gimmicky, but I have seen external scripts that do worse with less code. If I recall correctly, this was a request that came up during the team meetings, but I didn't include it on the 1.0 milestone. I would suggest removing it from this patch, and letting someone else work on it if they want to see it included. It might be nice to paste this working code snippet into the issue though.

davehunt commented:
As you're taking care of a number of things here, could you also add in #14? It occured to me that it would be a nice feature, and trivial to implement.

bobsilverberg commented:
@davehunt I have addressed your comments above in the latest commit. r?

bobsilverberg commented:
I have also pushed these changes to the review app

davehunt commented on scripts/webqa.coffee:
```diff
@@ -0,0 +1,37 @@
+    if howSoon.length
+      meetingTime = howSoon
+    else
+      meetingTime = " at 9AM Pacific every Thursday"
```
Can we use:

```coffeescript
meeting = (res, when) ->
  when ?= 'at 9am Pacific every Thursday'
```

This is really just a style nit, you can chose to ignore it. :)

davehunt commented on scripts/webqa.coffee:
```diff
@@ -0,0 +1,37 @@
+    res.send "For information on Vidyo: https://wiki.mozilla.org/Vidyo"
+
+  # announce or provide details of our team meeting
+  robot.respond /meeting(.*)/i, (res) ->
```
I'd prefer if we excluded any whitespace between the command and the argument. Something like `meeting([\s]+(.*))` should work.

bobsilverberg commented:
Thanks for all your help, @davehunt. I'm shipping it!
